plugins/newsdownloader.koplugin/epubdownloadbackend.lua

```diff
@@ -24,6 +24,48 @@ local max_redirects = 5; --prevent infinite redirects
 local TIMEOUT_CODE = "timeout" -- from socket.lua
 local MAXTIME_CODE = "maxtime reached" -- from sink_table_with_maxtime
 
+-- filter HTML using CSS selector
+local function filter(text, element)
+    local htmlparser = require("htmlparser")
+    local root = htmlparser.parse(text)
+    local filtered = ""
+    local selectors = {
+        "main",
+        "article",
+        "div#main",
+        "#main-article",
+        ".main-content",
+        "#body",
+        "#content",
+        ".content",
+        "div#article",
+        "div.article",
+        "div.post",
+        "div.post-outer",
+        ".l-root",
+        ".content-container",
+        ".StandardArticleBody_body",
+        "div#article-inner",
+        "div#newsstorytext",
+        "div.general",
+        }
+    if element then selectors[0] = element end
+    for _, sel in pairs(selectors) do
```

Comment:
There's a thing in Lua with table-tables and array-tables - and array tables start at 1.
So, this is not super proper (but we have some occurences of that kind of stuff in the code, and it often caused us some headaches :), so it's best to do that right.
Your `selectors` table is defined as an array of 18 elements: 1 .. 18.

You shouldn't create `selectors[0]`  as 0 is not an index (Lua starts at 1) - you're creating a `key=0 / value=element`  item in the table-table (one Lua table can be both a table-table and an array-table !).
Best to use `table.insert(selector, 1, element)` instead.

Then, using `for k, v = pairs(selectors)`, you iterate the table as a k/v table, and you get the keys in any order (may be that does not bother you, but if there's some sense of priority in the selectors list, it is lost).
Best to use `for i, v = ipairs(selectors)` (_ipairs_ with a _i_ like _iterate_) to iterate thru 1..18 (your selectors[0] would be ignored there).

Reply:
Huh, that is why it worked, and I wasn't made aware why does it not work. Thanks.